### PR TITLE
Add transactional support for ENR temp tables

### PR DESCRIPTION
**Pull request by time-change**
### Description

ENR temp tables were not sensitive to transactional behavior. We add it here by introducing a new structure which will track uncommitted ENR tuples and process them accordingly, depending on if the active transaction is committed or rolled back. 

Tests are added in babelfish-for-postgresql/babelfish_extensions#2522
 
### Issues Resolved

BABEL-4864
 
### Check List

- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/access/transam/xact.c [hunk 1]
```diff
@@ -380,6 +380,10 @@ IsTransactionState(void)
 {
 	TransactionState s = CurrentTransactionState;
 
+	/* During ENR Rollback, we may need to call relation_open during TRANS_ABORT. */
```
Comment (surendravishnoi) [inline src/backend/access/transam/xact.c]: comment clearly says that TRANS_ABORT is unsafe. How do we reach to conclusion that it is safe for our use case?

File: src/backend/access/transam/xact.c [hunk 2]
```diff
@@ -631,7 +635,10 @@ AssignTransactionId(TransactionState s)
 
 	/* Assert that caller didn't screw up */
 	Assert(!FullTransactionIdIsValid(s->fullTransactionId));
-	Assert(s->state == TRANS_INPROGRESS);
+
+	/* In ENR Rollback, we may need to call relation_open during TRANS_ABORT. */
+	if (temp_table_xact_support && !ENRInRollback() && s->state != TRANS_ABORT)
```
Comment (surendravishnoi) [inline src/backend/access/transam/xact.c]: this is very risky. There must be a reason for not assigning new transaction ids while transaction is in aborted state

File: src/backend/utils/misc/queryenvironment.c [hunk 3]
```diff
@@ -146,6 +145,12 @@ register_ENR(QueryEnvironment *queryEnv, EphemeralNamedRelation enr)
 	Assert(enr != NULL);
 	Assert(get_ENR(queryEnv, enr->md.name, false) == NULL);
 
+	if (enr->md.name[0] != '#')
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: It is not safe to assume that anything not starting with # must be a table variable? or we are just tracking all non temp table objects? if second, please use appropriate name for the same

File: src/backend/utils/misc/queryenvironment.c [hunk 4]
```diff
@@ -669,13 +674,129 @@ find_enr(Form_pg_depend entry)
...
+		default:
+			/* Shouldn't reach here */
+			elog(ERROR, "Did not find matching ENR catalog entry for catoid %d", catoid);
+			break;
+	}
+
+	/*
+	 * Extra UPDATEs can be skipped because we don't care about intermediate updates - only the initial state.
+	 */
+	if (op == ENR_OP_UPDATE && *list_ptr != NIL)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: how will we handle sub transaction in such cases? even though an update is intermediate in transaction but could be first one in current sub transaction

File: src/backend/utils/misc/queryenvironment.c [hunk 5]
```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
...
+ * In other words, search through the list of uncommitted tuples for a match, and return if true.
+ */
+bool
+ENRTupleIsDropped(Relation rel, HeapTuple tup)
+{
+	Oid catalog_oid;
+	ListCell *lc;
+	QueryEnvironment *queryEnv = currentQueryEnv;
+
+	if (sql_dialect != SQL_DIALECT_TSQL)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: why are we relying on just dialect here? shouldn't we look into what kind of object it is?

File: src/backend/utils/misc/queryenvironment.c [hunk 6]
```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
...
+		ListCell *lc2;
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		if (enr->md.enrtype != ENR_TSQL_TEMP)
+			continue;
+
+		/* Obtain oid to search by from tuple*/
+		switch (catalog_oid)
+		{
+			case RelationRelationId:
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: this is code duplication with existing logic, may be we can refactor to return tuple and caller can do relevant operation later

File: src/backend/utils/misc/queryenvironment.c [hunk 7]
```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
...
+}
+
+/*
+ * ENRCommitChanges
+ *
+ * 1. Free uncommitted tuple data from each ENR, then mark it as committed.
+ * 2. Finish freeing ENRs that were dropped in this transaction.
+ */
+void
+ENRCommitChanges(QueryEnvironment *queryEnv)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c:1120]: what if there a failure after this point? would a rollback happen and how will we do undo in that case?

File: src/backend/utils/misc/queryenvironment.c [hunk 8]
```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
...
+	/* Add back any ENRs that must be restored. */
+	foreach (lc, queryEnv->dropped_namedRelList)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		Assert(enr->md.enrtype == ENR_TSQL_TEMP);
+		Assert(!enr->md.is_table_variable);
+
+		/* If this was dropped in the current subtransaction, then we must restore it. */
+		if (enr->md.dropped_subid != InvalidSubTransactionId && enr->md.dropped_subid >= subid) // 4
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c:1306]: how is dropped subid work here? we have 1, 2... 10 drops in 10 savepoints, how will we know up to which point we want to cleanup with single dropped subid?

File: src/backend/utils/misc/queryenvironment.c [hunk 9]
```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
...
+	 */
+	foreach(lc, queryEnv->namedRelList)
+	{
+		Relation rel;
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		if (enr->md.enrtype != ENR_TSQL_TEMP || enr->md.is_table_variable)
+			continue;
+
+		if (!enr->md.is_committed && enr->md.created_subid >= subid)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c:1324]: same as above, how will single sub id let us know which ENRs to drop?

File: src/backend/utils/misc/queryenvironment.c [hunk 10]
```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
...
+			List *uncommitted_cattups = enr->md.uncommitted_cattups[i];
+			ListCell *lc2;
+
+			foreach(lc2, uncommitted_cattups)
+			{
+				ENRUncommittedTuple uncommitted_tup = (ENRUncommittedTuple) lfirst(lc2);
+				bool skip_cache_inval = false;
+
+				if (!OidIsValid(uncommitted_tup->catalog_oid) || !HeapTupleIsValid(uncommitted_tup->tup))
+					elog(ERROR, "ROLLBACK failed! Invalid temp table entries were found.");
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: what happens when we encounter this error? a rollback cannot fail

File: src/backend/utils/misc/queryenvironment.c [hunk 11]
```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
...
+
+	foreach(lc, queryEnv->namedRelList)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		if (enr->md.enrtype != ENR_TSQL_TEMP || enr->md.is_table_variable)
+			continue;
+
+		ENRDeleteUncommittedTupleData(InvalidSubTransactionId, enr);
+		enr->md.is_committed = true;
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c:1138]: we mark it committed here, what happens when a new transaction starts?

File: src/backend/utils/misc/queryenvironment.c [hunk 12]
```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
...
+		if (enr->md.enrtype != ENR_TSQL_TEMP || enr->md.is_table_variable)
+			continue;
+
+		if (!enr->md.is_committed)
+		{
+			enrs_to_drop = lappend(enrs_to_drop, enr);
+			continue;
+		}
+
+		/* ROLLBACK block. */
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: what if there are multiple table versions all committed? like
begin;
create; drop; create; drop; create; drop; create;
commit
why are we keeping all versions and not just one?

File: src/backend/utils/misc/queryenvironment.c (hunk 3 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: it is the second, I will change the name

File: src/backend/access/transam/xact.c (hunk 2 above)
Comment (time-change) [reply src/backend/access/transam/xact.c]: See discussion above

File: src/backend/access/transam/xact.c (hunk 1 above)
Comment (time-change) [reply src/backend/access/transam/xact.c]: Moved them to AbortTransaction, right before we set `s->state = TRANS_ABORT`. Functionally this is similar, as the Xact callbacks are invoked only and directly from AbortTransaction/AbortSubTransaction. The difference here is that, this code can be invoked regardless of whether our xact_callback functions are installed.

So if it seems necessary and this approach works, we could introduce a bool in _PG_init where the xact callbacks are registered so that the logic remains consistent. For ex, I'm not sure whether the subxact callbacks get removed if Babelfish is uninstalled from an instance, but this code would remain there regardless.

File: src/backend/utils/misc/queryenvironment.c (hunk 4 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: That's true, with subtransaction support this is no longer true, so removed this line to track intermediate updates.

File: src/backend/utils/misc/queryenvironment.c (hunk 5 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: ENRTupleIsDropped is only called from ENRdropTuple - so we don't have information other than the target Relation and HeapTuple, I'm not sure what else we can check for at this point in the function

File: src/backend/utils/misc/queryenvironment.c (hunk 11 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:1138]: If it is committed, then it was created outside of the current transaction. We track whether an enr is committed so that during ROLLBACK, we can be sure to drop any ENRs created in the current transaction.

File: src/backend/utils/misc/queryenvironment.c (hunk 12 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: I don't follow the question here - this is in the ROLLBACK block. I'm not sure how multiple table versions can be committed without running into name collision.

File: src/backend/utils/misc/queryenvironment.c (hunk 8 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:1306]: There is no way for an ENR to have its dropped subid altered. If the ENR is dropped in a subtransaction, then the subxactid is recorded, and the ENR is moved into the dropped_relid list - this is effectively the same as actually dropping it, but the information sticks around until commit/rollback. If the ENR is later recreated (ie, same name), a **new ENR is created**. Then if that new ENR is dropped, it is moved into the dropped_relid list with the new subtransaction id in which it was created/dropped.

Otherwise, if the same ENR is added back to namedRelList as part of rollback, you can see that the dropped_subid is reset. Then if it is dropped in a later subtransaction, it will be updated accordingly. 

Let me know if I misunderstood the question

File: src/backend/utils/misc/queryenvironment.c (hunk 9 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:1324]: See comment above

File: src/backend/utils/misc/queryenvironment.c (hunk 10 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: It should not fail, because this would mean a violated assumption or some issue with memory. But I thought it wouldn't be appropriate to emit a FATAL error to close the server since we are dealing with temp objects here - any bad state could be forcibly cleaned up by opening a new session.

File: src/backend/utils/misc/queryenvironment.c (hunk 7 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:1120]: If a failure happens at this point, a rollback would be called and changes in the transaction would be rolled back as usual. 

If any sort of failure happens during ENRRollback though, we would get a crash as it's called during AbortTransaction.

File: src/backend/utils/misc/queryenvironment.c (hunk 10 above)
Comment (lejaokri) [reply src/backend/utils/misc/queryenvironment.c]: I think we will crash anyway.

For instance, try setting `uncommitted_tup->catalog_oid=0` then the code goes to error codepath over and over again until we hit a maximum.

ie
```
ereport(ERROR) -> triggers error handling
error handling goes to AbortTransaction
AbortTransaction calls ENRRollbackChanges
```

File: src/backend/utils/misc/queryenvironment.c (hunk 3 above)
Comment (surendravishnoi) [reply src/backend/utils/misc/queryenvironment.c]: also, can we assume anything starting with # must be a temp table? Please validate that trigger inserted/deleted tables cannot start with #

File: src/backend/utils/misc/queryenvironment.c [hunk 13]
```diff
@@ -146,6 +145,12 @@ register_ENR(QueryEnvironment *queryEnv, EphemeralNamedRelation enr)
 	Assert(enr != NULL);
 	Assert(get_ENR(queryEnv, enr->md.name, false) == NULL);
 
+	if (enr->md.name[0] != '#')
+		enr->md.is_not_bbf_temp_table = true;
+
+	if (!enr->md.is_not_bbf_temp_table && GetCurrentSubTransactionId() != InvalidSubTransactionId)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: why not call it is_bbf_temp_table? !not seems non-intuitive

File: src/backend/utils/misc/queryenvironment.c [hunk 14]
```diff
@@ -669,13 +674,123 @@ find_enr(Form_pg_depend entry)
...
+static void
+ENRAddUncommittedTupleData(EphemeralNamedRelation enr, Oid catoid, ENRTupleOperationType op, HeapTuple tup)
+{
+	ENRUncommittedTuple uncommitted_tup;
+	List **list_ptr = NULL;
+	
+	if (in_enr_rollback)
+		return;
+	
+	uncommitted_tup = (ENRUncommittedTuple) palloc0(sizeof(ENRUncommittedTupleData));
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c:687]: what memory context we are using here? we need to make sure is at transaction level

File: src/backend/utils/misc/queryenvironment.c [hunk 15]
```diff
@@ -669,13 +674,123 @@ find_enr(Form_pg_depend entry)
...
+		foreach(lc, enr->md.uncommitted_cattups[i])
+		{
+			ENRUncommittedTuple uncommitted_tup = (ENRUncommittedTuple) lfirst(lc);
+
+			if (subid != InvalidSubTransactionId && uncommitted_tup->subid < subid)
+				continue;
+
+			heap_freetuple(uncommitted_tup->tup);
+			pfree(uncommitted_tup);
+			if (uncommitted_tup->subid >= subid)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: this check is unnecessary

File: src/backend/utils/misc/queryenvironment.c (hunk 6 above)
Comment (surendravishnoi) [reply src/backend/utils/misc/queryenvironment.c]: can we refactor it?

File: src/backend/utils/misc/queryenvironment.c (hunk 7 above)
Comment (surendravishnoi) [reply src/backend/utils/misc/queryenvironment.c:1120]: Point is that we have already got rid of undo data from ENR here. So even though ENRollback is called, it has no undo data to do actual rollback

File: src/backend/utils/misc/queryenvironment.c [hunk 16]
```diff
@@ -978,8 +1105,469 @@ void ENRDropEntry(Oid id)
...
+		return false;
+
+	catalog_oid = RelationGetRelid(rel);
+
+	foreach(lc, queryEnv->namedRelList)
+	{
+		ListCell *lc2;
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		if (enr->md.enrtype != ENR_TSQL_TEMP)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: why are we not checking for is_not_bbf_temp_table here?

File: src/backend/utils/misc/queryenvironment.c [hunk 17]
```diff
@@ -669,13 +674,123 @@ find_enr(Form_pg_depend entry)
...
+				continue;
+
+			heap_freetuple(uncommitted_tup->tup);
+			pfree(uncommitted_tup);
+			if (uncommitted_tup->subid >= subid)
+				enr->md.uncommitted_cattups[i] = foreach_delete_current(enr->md.uncommitted_cattups[i], lc);
+		}
+		if (subid == InvalidSubTransactionId)
+		{
+			list_free(enr->md.uncommitted_cattups[i]);
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c:773]: will list free take care of all elements inside list?

File: src/backend/utils/misc/queryenvironment.c (hunk 11 above)
Comment (surendravishnoi) [reply src/backend/utils/misc/queryenvironment.c:1138]: got it

File: src/backend/utils/misc/queryenvironment.c [hunk 18]
```diff
@@ -951,15 +1073,20 @@ bool ENRaddTuple(Relation rel, HeapTuple tup)
  */
 bool ENRdropTuple(Relation rel, HeapTuple tup)
 {
-	return _ENR_tuple_operation(rel, tup, ENR_OP_DROP);
+	/*
+	 * If we've already dropped it in this transaction but haven't committed, pretend the drop is done. 
+	 */
+	if (temp_table_xact_support && ENRTupleIsDropped(rel, tup))
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: which use case is this? why would we call dropped on an already dropped tuple?

File: src/backend/utils/misc/queryenvironment.c (hunk 12 above)
Comment (surendravishnoi) [reply src/backend/utils/misc/queryenvironment.c]: in above example, we will have one entry in named rel list and 3 entries in dropped named list. We are transferring all dropped entries into named list. Why are we doing that? we just need the first one from the list which is committed, others can be dropped immediately

File: src/backend/utils/misc/queryenvironment.c [hunk 19]
```diff
@@ -978,8 +1105,469 @@ void ENRDropEntry(Oid id)
...
+			List *uncommitted_cattups = enr->md.uncommitted_cattups[i];
+			ListCell *lc2;
+
+			foreach(lc2, uncommitted_cattups)
+			{
+				ENRUncommittedTuple uncommitted_tup = (ENRUncommittedTuple) lfirst(lc2);
+				bool skip_cache_inval = false;
+
+				if (!OidIsValid(uncommitted_tup->catalog_oid) || !HeapTupleIsValid(uncommitted_tup->tup))
+					elog(ERROR, "ROLLBACK failed! Invalid temp table entries were found.");
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: this will result in server crash.

File: src/backend/utils/misc/queryenvironment.c [hunk 20]
```diff
@@ -978,8 +1105,469 @@ void ENRDropEntry(Oid id)
...
+				ENRUncommittedTuple uncommitted_tup = (ENRUncommittedTuple) lfirst(lc2);
+				bool skip_cache_inval = false;
+
+				if (!OidIsValid(uncommitted_tup->catalog_oid) || !HeapTupleIsValid(uncommitted_tup->tup))
+					elog(ERROR, "ROLLBACK failed! Invalid temp table entries were found.");	
+
+				if (uncommitted_tup->subid < subid)
+					continue;
+
+				if (uncommitted_tup->catalog_oid == IndexRelationId && uncommitted_tup->optype == ENR_OP_ADD)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: this code can be refactored and shared with rollback handling

File: src/backend/utils/misc/queryenvironment.c (hunk 3 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: yes, the trigger table names "inserted" and "deleted" are hardcoded in

File: src/backend/utils/misc/queryenvironment.c (hunk 14 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:687]: this is in queryEnv->memCtx, it is called in _ENR_tuple_operation. same ctx as Rollback/Commit

File: src/backend/utils/misc/queryenvironment.c (hunk 17 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:773]: list_free takes care of the list and all of the ListCell pointers in the list; the elements inside the list were freed in the previous loop

File: src/backend/utils/misc/queryenvironment.c (hunk 7 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:1120]: Yes, then functionally it would be a noop. I don't think theres anything wrong with that, for ex. an empty transaction block is fine: 

```
BEGIN transaction
ROLLBACK
```

File: src/backend/utils/misc/queryenvironment.c (hunk 12 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: ack, added a fix in ENRDropEntry

File: src/backend/utils/misc/queryenvironment.c (hunk 10 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: Modified this to just continue - that way we can still attempt to clean up as much as possible and avoid infinite looping in error codepath.

File: src/backend/access/transam/xact.c [hunk 21]
```diff
@@ -631,6 +631,7 @@ AssignTransactionId(TransactionState s)
 
 	/* Assert that caller didn't screw up */
 	Assert(!FullTransactionIdIsValid(s->fullTransactionId));
+
```
Comment (surendravishnoi) [inline src/backend/access/transam/xact.c]: avoid unnecessary changes

File: src/backend/utils/misc/queryenvironment.c (hunk 7 above)
Comment (surendravishnoi) [reply src/backend/utils/misc/queryenvironment.c:1120]: I am talking about this case
begin;
add ENR
commit -> there is failure after we delete uncommitted data, now rollback will happen. As part of rollback we need to remove ENR but we cannot as we have removed corresponding uncommitted data in failed commit

File: src/backend/utils/misc/queryenvironment.c [hunk 22]
```diff
@@ -978,8 +1103,287 @@ void ENRDropEntry(Oid id)
...
+		
+		Form_pg_index idx_form = (Form_pg_index) GETSTRUCT(uncommitted_tup->tup);
+		EphemeralNamedRelation tmp_enr = get_ENR_withoid(queryEnv, idx_form->indrelid, ENR_TSQL_TEMP);
+		if (tmp_enr)
+		{
+			skip_cache_inval = true;
+		}
+	}
+
+	rel = relation_open(uncommitted_tup->catalog_oid, AccessExclusiveLock);
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: do we really need to ope relation in exclusive mode?

File: src/backend/utils/misc/queryenvironment.c [hunk 23]
```diff
@@ -669,13 +674,123 @@ find_enr(Form_pg_depend entry)
 	return NULL;
 }
 
+bool 
+ENRInRollback()
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: where do we use this function?

File: src/backend/utils/misc/queryenvironment.c [hunk 24]
```diff
@@ -978,8 +1103,287 @@ void ENRDropEntry(Oid id)
...
+			continue;
+
+		if (!enr->md.is_committed)
+		{
+			enrs_to_drop = lappend(enrs_to_drop, enr);
+			continue;
+		}
+
+		/* ROLLBACK block. */
+		in_enr_rollback = true;
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: instead of setting variable, preferably pass a bool indicating the same to skip undo during rollback

File: src/backend/utils/misc/queryenvironment.c [hunk 25]
```diff
@@ -978,8 +1103,287 @@ void ENRDropEntry(Oid id)
...
+	/* If we are dropping a committed ENR, wait until COMMIT to free it. */
+	if (temp_table_xact_support && enr->md.is_bbf_temp_table)
+	{
+		ListCell *lc;
+		enr->md.dropped_subid = GetCurrentSubTransactionId();
+		foreach(lc, currentQueryEnv->dropped_namedRelList)
+		{
+			EphemeralNamedRelation tmp_enr = (EphemeralNamedRelation) (lc);
+
+			if (strcmp(enr->md.name, tmp_enr->md.name) == 0)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: this is incorrect, we remove old entries only if they are in same transaction/sub-transaction

File: src/backend/utils/misc/queryenvironment.c (hunk 12 above)
Comment (surendravishnoi) [reply src/backend/utils/misc/queryenvironment.c]: my suggestion was to avoid adding them to named list and rather drop directly here. Otherwise, we are first adding them to name list, later putting them into ends_to_drop list and drop later

File: src/backend/utils/misc/queryenvironment.c (hunk 7 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:1120]: So it is true that if there is a failure after this point that requires a rollback, we won't be able to rollback the metadata properly.

However, I don't see a clean solution for this. If we leave the data around in case it needs to be rolled back AFTER this point in the commit, well then it still needs to be cleaned up at some point. And then the question can be raised - what happens if we hit an error after the cleanup happens then? IMO, it is alright to leave this edge case around because:

1) If we have a failure after ENRDeleteUncommittedTupleData, in our code it could only be during the pfree calls to queryEnv->dropped_namedRelList. If those calls fail for some reason, it would be a memory issue that causes server restart, so the bad metadata would be cleaned up. 

2) If we have a failure after ENRDeleteUncommittedTupleData but it is NOT in our code, then it is in Postgres native transaction handling code, (ie. some point in CommitTransaction in xact.c). I am not sure what PG would do in that scenario as well, since there are a few calls before the section where it is "too late" to abort a transaction inside CommitTransaction. 

3) Ultimately if that does somehow happen, the customer will be able to reach a clean slate by opening a new session.

File: src/backend/utils/misc/queryenvironment.c (hunk 12 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: The original example you gave was 

begin
create; drop; create; drop; create; drop; create
commit.

so the optimization works here where we won't track those intermediate dropped ENRs in dropped_namedRelList.

but with ROLLBACK, i think the overhead is minimal. since none of these are marked as committed, the code will skip processing and then add them to the dropped ENR list anyways. so the difference is just dropping them slightly earlier in this call instead of dropping them at the end.

File: src/backend/utils/misc/queryenvironment.c [hunk 26]
```diff
@@ -978,8 +1095,279 @@ void ENRDropEntry(Oid id)
...
+	/* If we are dropping a committed ENR, wait until COMMIT to free it. */
+	if (temp_table_xact_support && enr->md.is_bbf_temp_table)
+	{
+		ListCell *lc;
+		enr->md.dropped_subid = GetCurrentSubTransactionId();
+		foreach(lc, currentQueryEnv->dropped_namedRelList)
+		{
+			EphemeralNamedRelation tmp_enr = (EphemeralNamedRelation) (lc);
+
+			if (strcmp(enr->md.name, tmp_enr->md.name) == 0 && tmp_enr->md.dropped_subid == enr->md.dropped_subid)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: given that we are handling this during rollback itself, we can skip this optimisation in favour of simplicity

**Review by surendravishnoi — DISMISSED** [review]

File: src/backend/utils/misc/queryenvironment.c (hunk 26 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c]: Ack, fixed. Noting that surendra has approved PR (in case this gets lost due to force push).

**Review by lejaokri — APPROVED** [review]
Looking at notes seems Surendra has approved this one.

**Review by forestkeeper — APPROVED** [review]
